packages/ActionBar/src/hooks/useColumnsArrangement/useColumnsArrangement.js

```diff
@@ -31,23 +32,71 @@ function updateHiddenColumnIdFromLocalStorage(localStoragePrefix, columnId, isVi
   }
 }
 
+function hasDifferences(originalArray, newArray) {
+  if (originalArray.length !== newArray.length) return true;
+
+  return difference(originalArray, newArray).length > 0 || difference(newArray, originalArray).length > 0;
+}
+
+function getInitialOrder({ defaultOrder, isLocalStorageEnabled, localStoragePrefix }) {
+  if (!isLocalStorageEnabled) return defaultOrder;
+
+  const cachedOrder = getColumnIdsfromLocalStorage(
+    getLocalStorageKey(localStoragePrefix, savedItem.ORDER),
+    defaultOrder
+  );
+
+  if (hasDifferences(cachedOrder, defaultOrder)) {
+    window.localStorage.setItem(getLocalStorageKey(localStoragePrefix, savedItem.ORDER), JSON.stringify(defaultOrder));
+    return defaultOrder;
+  }
+
+  return cachedOrder;
+}
+
+function getInitialHiddenColumnIds({
+  defaultHiddenColumnIds,
+  defaultOrder,
+  isLocalStorageEnabled,
+  localStoragePrefix,
+}) {
+  if (!isLocalStorageEnabled) return new Set(defaultHiddenColumnIds);
+
+  const cachedHiddenColumns = getColumnIdsfromLocalStorage(
+    getLocalStorageKey(localStoragePrefix, savedItem.VISIBILITY),
+    defaultHiddenColumnIds
+  );
+
+  const updatedHiddenColumns = cachedHiddenColumns.filter(columnId => defaultOrder.includes(columnId));
+
+  if (hasDifferences(updatedHiddenColumns, cachedHiddenColumns)) {
+    window.localStorage.setItem(
+      getLocalStorageKey(localStoragePrefix, savedItem.VISIBILITY),
+      JSON.stringify(updatedHiddenColumns)
+    );
+  }
+
+  return new Set(updatedHiddenColumns);
+}
+
 export default function useColumnsArrangement({
   defaultOrderedColumnIds: defaultOrder,
   disabledColumnIds = [],
   defaultHiddenColumnIds = [],
   localStoragePrefix = null,
 }) {
   const isLocalStorageEnabled = localStoragePrefix !== null;
-  const [order, setOrder] = React.useState(defaultOrder);
-  const [hiddenColumnIds, setHiddenColumnIds] = React.useState(() =>
-    isLocalStorageEnabled
-      ? new Set(
-          getColumnIdsfromLocalStorage(
-            getLocalStorageKey(localStoragePrefix, savedItem.VISIBILITY),
-            defaultHiddenColumnIds
-          )
-        )
-      : new Set(defaultHiddenColumnIds)
+  const [order, setOrder] = React.useState(
+    getInitialOrder({ defaultOrder, isLocalStorageEnabled, localStoragePrefix })
```

Comment:
I think you can have:

```
  const [order, setOrder] = React.useState(() => {
   if (!isLocalStorageEnabled) return defaultOrder;
   return getInitialOrder({ defaultOrder localStoragePrefix }) 
  })
```

and ignore not need to pass down `isLocalStorageEnabled` ?